### PR TITLE
[WIP] Witness matching

### PR DESCRIPTION
**Pull request by DougGregor**
Work-in-progress to make witness matching in Sema sane, particularly when working with generic requirements. This kills off a hack in the constraint solver and starts to record more information about witnesses in protocol conformances, so SILGen doesn't need to rederive the same information.

This is still very, very busted. The Sema part seems to be working properly (no failures in test/Generics or test/Constraints, where we have the interesting matching cases), but there are still 11 failures in SILGen. The current concern is that we're dropping information for "default witnesses", which is manifesting as some of the failures. Current failing test:

``` swift
protocol P2 {
  associatedtype A
  func f1(_ a: A)
}

extension P2 {
  func f1(_ a: A) { }
}
```

### CODE REVIEW

**Comment by jckarter** [comment]
The default witness stuff is @slavapestov's. Maybe he has thoughts.

**Comment by DougGregor** [comment]
I think I know what's wrong with default witnesses. SILGen has some special magic for them, but it shouldn't need to. Sema seems to be computing the right information already.

File: stdlib/public/core/Mirror.swift [hunk 1]
```diff
@@ -289,9 +289,10 @@ public struct Mirror {
     self._makeSuperclassMirror = Mirror._superclassIterator(
       subject, ancestorRepresentation)
       
-    self.children = Children(
+    let lazyChildren =
       unlabeledChildren.lazy.map { Child(label: nil, value: $0) }
-    )
+    self.children = Children(lazyChildren)
```
Comment (slavapestov) [inline stdlib/public/core/Mirror.swift]: What happened here?

File: include/swift/AST/Decl.h [hunk 2]
```diff
@@ -3333,7 +3333,7 @@ class ProtocolDecl : public NominalTypeDecl {
 
   ArrayRef<ProtocolDecl *> InheritedProtocols;
 
-  llvm::DenseMap<ValueDecl *, ConcreteDeclRef> DefaultWitnesses;
+  llvm::DenseMap<ValueDecl *, ValueDecl *> DefaultWitnesses;
```
Comment (slavapestov) [inline include/swift/AST/Decl.h]: Don't default witnesses require substitutions sometimes?

Imagine the requirement looks like:

    protocol P { func foo(x: Int) }

And the default witness is

    extension P { func foo<T>(x: T) { ... } }

File: lib/Sema/TypeCheckProtocol.cpp [hunk 3]
```diff
@@ -897,6 +902,202 @@ matchWitness(TypeChecker &tc,
...
+  // If there were no generic parameters, we're done.
+  if (allGenericParams.empty()) {
+    reqGenericSig = nullptr;
+    return nullptr;
+  }
+
+  // Next, add each of the requirements (mapped from the requirement's
+  // interface types into the abstract type parameters).
+  RequirementSource source(RequirementSource::Explicit, SourceLoc());
+  for (auto &reqReq : reqSig->getRequirements().slice(1)) {
```
Comment (slavapestov) [inline lib/Sema/TypeCheckProtocol.cpp]: Why are you skipping the first requirement?

File: include/swift/AST/Decl.h (hunk 2 above)
Comment (DougGregor) [inline include/swift/AST/Decl.h]: SILGen was trying to make up for this, but it shouldn't. Fix coming!

File: stdlib/public/core/Mirror.swift (hunk 1 above)
Comment (DougGregor) [inline stdlib/public/core/Mirror.swift]: That's from a totally unrelated commit that was sitting around in my local tree. It got merged and I didn't rebase.

File: lib/Sema/TypeCheckProtocol.cpp (hunk 3 above)
Comment (slavapestov) [inline lib/Sema/TypeCheckProtocol.cpp]: Isn't the first requirement the witness marker for Self, and the second requirement the conformance requirement for Self : Proto?

Comment (DougGregor) [inline lib/Sema/TypeCheckProtocol.cpp]: Ah, good point. There's probably no reason to drop anything... at worst, these are trivially-solved requirements.

File: test/SILGen/dependent_member_lowering.swift [hunk 4]
```diff
@@ -9,13 +9,13 @@ struct Foo<T>: P {
   typealias A = T.Type
 
   func f(_ t: T.Type) {}
-  // CHECK-LABEL: sil hidden [transparent] [thunk] @_TTWurGV25dependent_member_lowering3Foox_S_1PS_FS1_1{{.*}} : $@convention(witness_method) <T> (@in @thick T.Type, @in_guaranteed Foo<T>) -> ()
-  // CHECK:       bb0(%0 : $*@thick T.Type, %1 : $*Foo<T>):
+  // CHECK-LABEL: sil hidden [transparent] [thunk] @_TTWurGV25dependent_member_lowering3Foox_S_1PS_FS1_1{{.*}} : $@convention(witness_method) <τ_0_0> (@in @thick τ_0_0.Type, @in_guaranteed Foo<τ_0_0>) -> ()
+  // CHECK:       bb0(%0 : $*@thick τ_0_0.Type, %1 : $*Foo<τ_0_0>):
```
Comment (DougGregor) [inline test/SILGen/dependent_member_lowering.swift]: Well, we would have to at least drop the names on the type parameters coming from generic requirements (because they're moving to a new depth), and it felt weird to drop some but not all of them.

Comment (DougGregor) [inline test/SILGen/dependent_member_lowering.swift]: So, yeah, I'm not in favor of trying to keep the sugar here. Users don't care, and it's more clear (IMO!) to represent the parameters with depth/index.

**Comment by DougGregor** [comment]
@swift-ci please test

**Comment by DougGregor** [comment]
@swift-ci please benchmark

**Comment by DougGregor** [comment]
@swift-ci please test

**Comment by DougGregor** [comment]
@swift-ci please benchmark

**Comment by DougGregor** [comment]
@swift-ci please test Linux

**Comment by DougGregor** [comment]
Yay for functional progress, but disabling devirtualization of generic requirements (even temporarily) is a nonstarter

**Comment by DougGregor** [comment]
I believe I've fixed the devirtualizer now, so let's try again.

**Comment by DougGregor** [comment]
@swift-ci please benchmark

**Comment by DougGregor** [comment]
@swift-ci please test

**Comment by DougGregor** [comment]
Actually, I've messed with this PR so much that I've probably confused the PR testing infrastructure.
